### PR TITLE
Modernize npm

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Since NPM no longer allows us to download a JSON file with all packages. 

This implements the supported search API and removes a bunch of no longer needed code.

### CODE REVIEW

**Comment by billti** [comment]
Re _"This implements the supported search API"_, so what is the search experience now? Looks like you are removing most of the existing UI, so where does the user search, and where do the results appear? Can you post an [animated] screenshot showing the new search flow?

**Comment by paulvanbrenk** [comment]
I mostly removed stuff around the local NPM cache. The main search experience is still kept intact.

npminstall

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 1]
```diff
@@ -72,7 +72,7 @@
     <Reference Include="Microsoft.VisualStudio.Shell.Interop.14.0.DesignTime, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <EmbedInteropTypes>True</EmbedInteropTypes>
     </Reference>
-    <Reference Include="Microsoft.VisualStudio.Utilities, Version=$(VSTarget).0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
+    <Reference Include="Microsoft.VisualStudio.Utilities, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: Why switch to a hardcoded version? Isn't this going to be a problem going forward?

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 2]
```diff
@@ -804,25 +766,6 @@
...
-    </EmbeddedResource>
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.cs.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.de.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.es.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.fr.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.it.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.ja.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.ko.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.pl.resx" />
-    <EmbeddedResource Include="Options\NodejsNpmOptionsControl.pt-BR.resx" />
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: Have we tested this on an intl build yet?

File: Nodejs/Product/Nodejs/Nodejs.csproj (hunk 1 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: Plan is to move this to NPM packages, in the future... this is to make sure breaks happen at compile time, and not at run time.

File: Nodejs/Product/Nodejs/Nodejs.csproj (hunk 2 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: yes, the entire control is gone.

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                lock (this.queuelock)
+                {
+                    cmdr = this.npmController.CreateNpmCommander();
+
+                    this.commander = cmdr;
+                }
+
+                await cmdr.ExecuteNpmCommandAsync(info.Arguments);
+            }
+            finally
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: This function could interleave with itself and cause `commander` / `isExecutingCommand` to be out of sync with each other. Writes to those fields need to be in the same lock block.

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                        this.currentCommand = this.commandQueue.Dequeue();
+                    }
+                    else
+                    {
+                        this.currentCommand = null;
+                    }
+                    count = this.commandQueue.Count;
+                }
+
+                if (null != this.currentCommand)
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Swap ><

File: Nodejs/Product/Npm/PackageJsonFactory.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@ public class PackageJsonFactory
     {
         public static IPackageJson Create(IPackageJsonSource source)
         {
-            return null == source.Package ? null : new PackageJson(source.Package);
+            return source.Package == null ? null : new PackageJson(source.Package);
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: `??`

File: Nodejs/Product/Npm/SPI/NpmController.cs [hunk 6]
```diff
@@ -276,11 +236,11 @@ private FileSystemWatcher CreateModuleDirectoryWatcherIfDirectoryExists(string d
             return watcher;
         }
 
-        private void Watcher_Modified(object sender, FileSystemEventArgs e)
+        private void WatcherModified(object sender, FileSystemEventArgs e)
         {
             string path = e.FullPath;
 
-            // Check that the file is either a package.json file, or exists in the node_modules directory
+            // Check that the file is either a package.json file, or exists in the nodemodules directory
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Npm/SPI/NpmController.cs]: Revert

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                {
+                    Execute(this.currentCommand);
+                }
+            }
+        }
+
+        public async Task<IEnumerable<IPackage>> GetCatalogPackagesAsync(string filterText)
+        {
+            var relativeUri = string.Format("/-/v1/search?text={0}", filterText);
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Need to URL-encode the text component

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+            var searchUri = new Uri(defaultRegistryUri, relativeUri);
+
+            var request = WebRequest.Create(searchUri);
+            using (var response = await request.GetResponseAsync())
+            {
+                var reader = new StreamReader(response.GetResponseStream());
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Add a comment showing the expected format of the JSON so people can follow this later

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+            var pkgList = new List<IPackage>();
+
+            // Inside the array, each object is an NPM package
+            var builder = new NodeModuleBuilder();
+            while (jsonReader.Read())
+            {
+                switch (jsonReader.TokenType)
+                {
+                    case JsonToken.PropertyName:
+                        if (StringComparer.OrdinalIgnoreCase.Equals(jsonReader.Value, "package"))
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Ignoring case in JSON?

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+                    || info.Equals(this.currentCommand))
+                {
+                    return;
+                }
+                this.commandQueue.Enqueue(info);
+                Monitor.PulseAll(this.queuelock);
+            }
+        }
+
+        public void QueueCommand(string arguments)
```
Comment (RyanCavanaugh) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Where are the callers to this?

Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: In the `NpmPackageInstallViewModel`

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 3 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: redid the whole threading logic, should be sane now.

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 4 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: yep

File: Nodejs/Product/Npm/PackageJsonFactory.cs (hunk 5 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Npm/PackageJsonFactory.cs]: looks l did fix this up as I went

File: Nodejs/Product/Npm/SPI/NpmController.cs (hunk 6 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Npm/SPI/NpmController.cs]: done

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 7 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: done

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 8 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: added

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs (hunk 9 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: removed

**Pull request merged**